### ISSUE TITLE
Multi documents

### ISSUE THREAD

**Issue by rusimody** [toml-lang/toml#583]
In [yaml](http://yaml.org/spec/1.2/spec.html#id2760395) there is "---" to separate documents and "..." to end documents.  What if any is toml's position regarding multi-documents?

In particular how does one embed multiple toml docs into one stream?

**Comment by eksortso** [comment]
There's nothing right now in the standard to separate multiple documents in the same stream, although in #511 it was mentioned that the Hugo project uses `+++` to identify TOML documents in larger files. That's not part of the TOML standard, though. That's something that Hugo uses before parsing the TOML content.

Do you have a use case that would persuasively suggest that such a delimiter ought to be included in the standard?

**Comment by pocomane** [comment]
With https://github.com/toml-lang/toml/issues/581 you can use TOML to embend multiple TOMLs. Something like:

```
[[document]]
content = '''

[[fruit]]
name = 'apple'
[[fruit]]
name = 'orange'

'''

[[document]]
content = '''

[[color]]
name = 'red'
[[color]]
name = 'orange'

'''
```

Actually this system is already usable, but it becomes complicate if you have to embed documents with triple "'".  https://github.com/toml-lang/toml/issues/581 just simplifies this scenario.

We have still to count the logest "'" sequence in the embedded document, but probably we need something similar for any solution.

**Comment by eksortso** [comment]
Putting #581 aside for now, a TOML-strings-within-TOML method is a pattern in the same category as Hugo's front matter formats. It still requires an extra step before parsing to pull out the content.

Let me ask my question with more explicit detail:
Why should we add syntax to the TOML standard, that all compliant TOML parsers must acknowledge, that would enable said parsers to output any number of hash tables, not necessarily just a single hash table, from a single stream containing potentially multiple (completely independent) TOML documents?

It could be a nice feature that a TOML-using application could provide, but should it be part of the TOML standard, and hence part of all TOML parsers? Or can we leave it up to the applications to devise their own multi-doc standards, with the parsers predominantly handling just the contents from the multi-docs?

**Comment by rusimody** [comment]
On Wed, Dec 5, 2018 at 1:52 AM Dave Ostroske <notifications@github.com>
wrote:


> It could be a nice feature that a TOML-using application could provide,
> but should it be part of the TOML standard, and hence part of all TOML
> parsers?
>
Yes we know what happens when you try to be everything to everybody
(XML/yml etc)
Better to be focussed: eg html for documents, ini files (therefore toml)
for config; json for client-server data streaming etc
OTOH when the spec is too underspecified it tends to collect accretions of
random bolt-ons: think of the number of different ways that
quoting/escaping happens in csv files
So…
[From my side] I am not necessarily asking for a feature; just asking what
is the toml position on it

Putting #581 <https://github.com/toml-lang/toml/issues/581> aside for now,
> a TOML-strings-within-TOML method is a pattern in the same category as
> Hugo's front matter formats. It still requires an extra step before parsing
> to pull out the content.
>
> Let me ask my question with more explicit detail:
> Why should we add syntax to the TOML standard, that all compliant TOML
> parsers must acknowledge, that would enable said parsers to output any
> number of hash tables, not necessarily just a single hash table, from a
> single stream containing potentially multiple (completely independent) TOML
> documents?
>
> I can think of many levels. Here are some points in the spectrum between
the current status and full inclusion and implementation

   1. Current status (quo); ie say nothing implement nothing let each
   application that needs it do as it pleases
   2. Suggest "---" (or "+++" or whatever) with cautions/caveats but
   provide no implementation
   3. Choose one, and add as in yaml a load_all store_all implementation


Personally I'd appreciate at least a line or two in the docs such as
*Multiple toml documents as in yamls "---" is beyond the scope of toml;
toml is a config file format not a general data format; applications may do
as they please*

**Comment by pradyunsg** [comment]
Firstly, thanks for filing this issue!

And, thanks @eksortso for pitching in! Much appreciated. :)

---

The TOML specification does not have multi-document streams. This means that TOML parsers are expected to not implement it, because it's not part of the language. 

Digression: This is similar to expecting a C parser to not accept `print("Hello World")` as a valid program -- the intent and meaning of this text is likely clear to the reader but this syntax is just not part of the language.

> Current status (quo); ie say nothing implement nothing let each application that needs it do as it pleases

The applications that use TOML parser implementations, are free to extend them to allow additional syntax, to support additional functionality they may need -- the onus is on them to properly explicitly document their extension to the language. They're no longer using TOML but instead a language that is a superset of TOML.

> Suggest "---" (or "+++" or whatever) with cautions/caveats

No. The TOML specification has no business suggesting how to extend the language specified.

The caveats/trade-offs for choosing these are application specific - `+++` might be the choice in some cases, while in other cases `---` might be a better choice for the application at hand.

> Just asking what is the toml position on it

Given the cost of adding this syntax -- parser implementations would need to change (substantially in some cases) since a `.toml` file can now possibly be a list of hash tables -- the use cases presented aren't compelling enough to justify this additional complexity.

> I'd appreciate at least a line or two in the docs

This could be done, but I am wary of adding this -- I don't want to set precedent for mentioning features from other file formats that TOML hasn't adopted and, potentially, having to elaborate about it in the specification. Neither do I think we should add a "Why not X?" section in general. We already have a comparison with JSON and YAML which I think is sufficient.

---

@rusimody I think that should be enough to answer your questions and queries. Please feel free to comment if there's anything else.

---

If you want to make a case for adding document separators, please make a comment on #511.

**Comment by ChristianSi** [comment]
I agree that document separators/end-of-document markers should not become part of the spec, but I *do* think that it would be useful to have a FAQ section in the wiki, where the question "What do I do if I want to mark the end of a TOML document in a file/stream" (or similar) is answered, mentioning Hugo's example of using `+++` as one possibility.

The reason for this is exactly that the question *has* repeated been asked and will be asked again in the future. Hence it belongs into a FAQ, if only for the reason that this spares us from repeated issues being opened, discussed, and closed again in the issue tracker, which would be a waste of everyone's time.

**Comment by RMuskovets** [comment]
I think TOML is already good, if you want to split configs to multiple files, do sth like this:
```java
public TomlParseResult parseFile(String path) {...}

public TomlParseResult[] parseMultiple(String... path) {
    TomlParseResult res[] = new TomlParseResult[path.length];
    for (int i = 0; i < res.length; i++) {
        res[i] = parseFile(path[i]);
    }
    return res;
}
```

But if you wanna use TOML and not-TOML in one file, like this:
```xml
<root>
    <!-- TOML here -->
    <toml>
    some.toml = ""
    </toml>
    <xml>
    <test></test>
    </xml>
</root>
```
You should use parser for "bigger" file (XML in this example) and parse TOML from the XML using it (&lt;toml&gt; tag).

**Comment by pradyunsg** [comment]
> The reason for this is exactly that the question _has_ repeated been asked and will be asked again in the future.

Fair enough. I reckon a line to the "Comparison with Other Formats" regarding this, when contrasting with YAML should be sufficient.

**Comment by pradyunsg** [comment]
This got addressed in #636.
